### ISSUE TITLE
clippy::unnecessary_lazy_evaluations false positive when laziness prevents underflow

### ISSUE THREAD

**Issue by krtab** [rust-lang/rust-clippy#10071]
### Summary

I have the following code:

```rust
(pos % n != 0).then(|| pos - 1),
```

Because `pos` may be equal to 0, changing to `(pos % n != 0).then_some(pos - 1)` can trigger underflow and hence panic in debug mode. Nonetheless clippy warns:

```
warning: unnecessary closure used with `bool::then`
  --> src/bin/12.rs:31:13
   |
31 |             (pos % n != 0).then(|| pos - 1),
   |             ^^^^^^^^^^^^^^^----------------
   |                            |
   |                            help: use `then_some(..)` instead: `then_some(pos - 1)`
   |
   = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
```

### Lint Name

clippy::unnecessary_lazy_evaluations

### Reproducer

Cf summary

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by kraktus** [comment]
Hey, thanks for the report! This seem a duplicate of https://github.com/rust-lang/rust-clippy/issues/9422

**Comment by krtab** [comment]
Thanks! It is indeed. I'm closing it as such.
